src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -0,0 +1,257 @@
+            hir::ItemStruct(..) => {
+                let adt_def = self.tcx.adt_def(def_id);
+                for field_def in adt_def.all_fields() {
+                    let required_predicates: Vec<_> = required_predicates_to_be_wf(
+                        self.tcx,
+                        field_def,
+                        &local_required_predicates,
+                     ).into_iter().collect();
+                    local_required_predicates.insert(
+                        field_def.did,
+                        Rc::new(required_predicates),
+                    );
+                }
+            }
+            _ => {}
+        };
+
+        if local_required_predicates.len() > 0 {
+            self.changed = true;
+            self.inferred_outlives_map
+                .extend(local_required_predicates);
+        }
+    }
+
+    fn visit_trait_item(&mut self, trait_item: &'tcx hir::TraitItem) {}
+
+    fn visit_impl_item(&mut self, impl_item: &'tcx hir::ImplItem) {}
+}
+
+fn required_predicates_to_be_wf<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    field_def: &ty::FieldDef,
+    infered_outlives_map: &FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>>,
+) -> HashSet<ty::Predicate<'tcx>> {
+
+    let mut predicates = HashSet::new();
+    // Get the type of the field with identity substs applied.  For
+    // now, let's just see if that causes horrible cycles.
+    let field_ty = tcx.type_of(field_def.did);
+    match field_ty.sty {
+        // For each type `&'a T`, we require `T: 'a`
+        //
+        // If you have A = &'a T, then mt.ty here represents T and
+        // ty represents A (and r represents 'a). So we want T: 'a.
+        ty::TyRef(r, mt) => {
+            let outlives_pred = ty::Binder(ty::OutlivesPredicate(mt.ty, r))
+                .to_predicate();
+            predicates.insert(outlives_pred);
+        }
+
+        // For each struct/enum/union type `Foo<'a, T>`, we can
+        // load the current set of inferred and explicit predicates from
+        // `inferred_outlives_map` and see if those include `T: 'a`
+        ty::TyAdt(def, substs) => {
+
+            // Get type generics of def and lifetime(region) generics of def
+            let generics: &ty::Generics = tcx.generics_of(def.did);
+            // let gen_types: &Vec<ty::TypeParameterDef> = &generics.types;
+            // let gen_region: &Vec<ty::RegionParameterDef> = &generics.regions;
+            let gen_types: &HashSet<DefId> = &generics.types
+                .iter().map(|t| t.def_id).collect();
+            let gen_region: &HashSet<DefId> = &generics.regions
+                .iter().map(|t| t.def_id).collect();
+
+            // Iterate over all predicates in the infered_outlives_map.
+            // See if there are any OutlivesPredicate kind. If there
+            // exists a type and region that is also in this field
+            // then that should be returned.
+            for (_, p_list) in infered_outlives_map.iter() {
```

Comment:
So, we probably want to iterate over this:

```rust
let predicates =
    inferred_outlives
    .get(def.did)
    .map(|v| &v[..])
    .unwrap_or(&[]);
```

This would be the list of inferred predicates associated with the struct `Foo` (i.e., we are matching on the field type, and we see that it is `Foo`).

You can then go further by filtering those predicates down to the ones you care about:

```rust
let predicates = predicates.filter_map(|p| p.to_opt_type_outlives());
```

this will select only the `TypeOutlives` predicates.